zng/docs/zson.md

```diff
@@ -0,0 +1,683 @@
+* [2. The ZSON Data Model](#2-the-zson-data-model)
+  + [2.1 Primitive Types](#21-primitive-types)
+  + [2.2 Complex Types](#22-complex-types)
+  + [2.3 The Type Type](#23-the-type-type)
+  + [2.4 The Null Type](#24-the-null-type)
+  + [2.5 Type Definitions](#25-type-definitions)
+* [3. The ZSON Format](#3-the-zson-format)
+  + [3.1 Identifiers](#31-identifiers)
+  + [3.2 Type Decorators](#32-type-decorators)
+    - [3.2.1 Type Definitions](#321-type-definitions)
+  + [3.3 Primitive Values](#33-primitive-values)
+    - [3.3.1 String Escape Rules](#331-string-escape-rules)
+  + [3.4 Complex Values](#34-complex-values)
+    - [3.4.1 Record Value](#341-record-value)
+    - [3.4.2 Array Value](#342-array-value)
+    - [3.4.3 Set Value](#343-set-value)
+    - [3.4.4 Union Value](#344-union-value)
+    - [3.4.5 Enum Value](#345-enum-value)
+    - [3.4.6 Map Value](#346-map-value)
+  + [3.5 Type Value](#35-type-value)
+* [4. Examples](#4-examples)
+
+## 1. Introduction
+
+ZSON is a data model and serialization format that embodies both
+"structured" as well as "semi-structured" data.  It is
+semantically a superset of the schema-flexible document model of JSON
+and the schema-rigid table model of relational systems.
+
+ZSON has the look and feel of JSON but differs substantially
+under the covers as it enjoys:
+* a comprehensive, embedded type system with first-class types and definitions,
+* well-defined streaming semantics, and
+* a semantically-matched, performant binary form called ZNG.
+
+ZSON builds upon the elegant simplicity of JSON with "type decorators".
+These decorators appear throughout the value syntax unobstrusively,
+in an ergonomic and human-readable syntax, to establish a well-defined
+type of every value expressed in a ZSON input.
+
+In addition to the human-readable format described here, the ZSON data model
+is realized in an efficient binary format called
+[ZNG](spec.md) and a columnar variation
+of ZNG called [ZST](../../zst/README.md).
+While ZNG and ZST are suited for production workflows and operational systems
+where performance matters most,
+ZSON is appropriate for human-level inspection of raw data
+and for test and debug and for low-performance APIs where ergonomics matters
+more than performance.
+
+The ZSON design was inspired by the
+[Zeek TSV log format](https://docs.zeek.org/en/stable/examples/logs/)
+and [is semantically consistent with it](./zeek-compat.md).
+As far as we know, the Zeek log format pioneered the concept of
+embedding the schemas of log lines as metadata within the log files themselves
+and ZSON modernizes this original approach with a JSON-like syntax and
+binary format.
+
+### 1.1 Some Examples
+
+The simplest ZSON text is a single value, perhaps a string like this:
+```
+"hello, world"
+```
+There's no need for a type decorator here.  It's explicitly a string.
+
+A structured SQL table might look like this:
+```
+{ city: "Berkeley", state: "CA", population: 121643 (uint32) } (=schema)
```

Comment:
I suggest using a different type name for this example, since `schema` could cause some confusion with readers that are thinking about 'schemas' from other data models. (`cityinfo`?)

Reply:
Good point.  I had thought about that too.  It is ying and yang because it is supposed to be the schema of the table but there's nothing special about the word "schema" here.   I'll add some clarifying text in this regard and change the name to `city_schema`.